plugins/spike_remove.py

```diff
@@ -0,0 +1,312 @@
+        """
+        raw_spec_dat (numpy.array of shape C11YX)
+        returns:
+           corrected_data (numpy.array of shape C11YX)
+           pixel_corrected (int)
+           spikes corrected (int)
+        """
+        spikestep = self.threshold.value
+        specdat = numpy.squeeze(raw_spec_dat.copy())
+        assert specdat.ndim == 3
+        # this diff calculation requires higher numerical precision than 16 bits because it is squared.
+        # 32 uint should be good enough as the max diff < 2**16. However for the summation of ms_step it is more convenient to use float32.
+        # Maybe there is a trick to stick with uint32?
+        diffspec = numpy.diff(numpy.float32(specdat), axis=0) ** 2
+        size = numpy.shape(diffspec)
+        ms_step = (diffspec / numpy.prod(size)).sum()
+        
+        # We are now calculating the threshold based on the global average.
+        # Using a more local average could help identifying spikes
+        # more precisely although but it is more involved and possibly overkill
+        threshold = ms_step * spikestep**2     
+        # this sets sensitivity to spikes. Used in combination with this number
+        # is rather large now, we could be a bit more clever if a spike covers more pixels
+        spike_margin = 1  # number of pixels left and right of spike that are also corrected
+        spike_spacing = 3  # when spikes are considered to be two separate spikes
+        kCL = 0  # number of corrected pixels counter
+        kpix = 0 # spike counter
+    
+        # Look at each spectrum independently (as they were acquired independently)
+        for ii in range(size[1]):
+            for jj in range(size[2]):
+                specdiff = diffspec[:, ii, jj]
+                spec = specdat[:, ii, jj]
+                # These are the indices of the spike starts and ends.
+                spike_indices = numpy.argwhere(specdiff > threshold)
+                num_spike_indices = numpy.size(spike_indices)
+             
+                if num_spike_indices > 1: # only one step that deviates is no spike.
+                    kCL += 1
+                    spike_indices = numpy.squeeze(spike_indices)
+                    dif_spike = numpy.diff(spike_indices)
+
+                    # check whether there is more than one spike in the spectrum.
+                    # The first spike always starts at the first element of
+                    # spike_indices whereas the last spike always ends on the spike_indices
+                    spike_edges = numpy.argwhere(dif_spike > spike_spacing)
+                    # subtract 1 from the num_spike_indices to make it work for the last pixel
+                    spike_edges = numpy.append(spike_edges, num_spike_indices - 1)
+
+                    # number_of_spikes in spectrum
+                    nspikes_inspec = numpy.size(spike_edges)
+
+                    # distinguish spikes within spectrum
+                    for pp in range(nspikes_inspec):
```

Comment:
since `nspikes_inspec` is only used here you could also just write `for pp in range(numpy.size(spike_edges))`

Reply:
Indeed, and using `enumerate()` turns out to make the code look even nicer.